[RFC] channels: support buffered output and bytes sockets/stdio

bfredl opened the pull request:
Adds the following:

- [x] support bytes protocol also in socket and stdio channels
- [x] signal EOF in output handler, and add a mode where callback gets all output at once (with buffer limit?)
- [ ] ~~introspection of all channels and jobs~~ (do it in #6743)

bfredl commented:
Also change I would make, but in a separate PR: move almost all job/channel code from eval.c to channel.c, also expose functionality natively in lua, so lua functions/callbacks can use lua binary strings instead of linelists, also rpc functions can do directly lua<->msgpack (_primarily_ for semantics/convenience, only secondarily for performance). 

lua callbacks could even be run more asynchronously than vimL callbacks (i e in situations where vimL is forbidden) and must use `vim.schedule(callback)` for anything that touches vimL or mutates editor state. I'm not 100% we want this, but if we do, we should introduce this early on.

ZyX-I commented:
Lua is no more thread-safe then VimL is, except that you can asynchronously use different `lua_State`s and VimL has a bunch of globals scattered throught the code instead.

bfredl commented:
I'm not thinking of threads, just event handling in the main thread. There are many circumstances where the uv event loop is being run, but most events (that can execute arbitrary vimscript, most API functions) are blocked. We could be more liberal with executing lua callbacks if this behavior is introduced from start. Alternatively they can by default behave like vimL callbacks, and this behavior behind a flag.

bfredl committed:
channels: allow bytes sockets and stdio, and buffered bytes output

bfredl commented:
Now allows byte based sockets:

```
func! SockHandler(chan_id, data, event)
   let g:event = [a:chan_id, a:data, a:event]
endfunc

let x = sockconnect('pipe', '/some/nvim/socket`, {'on_output': 'SockHandler'})
call jobsend(x, "blargh")
echo g:event 
```
should see something like
`[4, ['~@l^A~@1~@^A~@^A~@^A~@^A~@^A~@^A~@^A~@^A~@n^A~@IMessage is not an array~@@'], 'socket']`
```

bfredl commented:
Also implement `stdioopen({'rpc':..., 'on_stdin':...})`. Only works in `--headless` mode, obviously. `nvim --embed` is then essentially a short-hand for `nvim --headless --cmd "call stdioopen({'rpc': v:true})"`. Also stdio channel is guaranteed to be 1, so tests can just use 1 instead of needing to query `nvim_get_api_info`

oni-link commented on src/nvim/channel.c:
```diff
@@ -0,0 +1,294 @@
+uint64_t channel_from_stdio(bool rpc, CallbackReader on_output, const char **error)
+{
+  if (!headless_mode) {
+    error = _("Stdio channel can only be opened in headless mode");
```
`*error` instead of `error`

oni-link commented on src/nvim/channel.c:
```diff
@@ -0,0 +1,294 @@
+  }
+
+  if (did_stdio) {
+    error = _("Stdio channel was already open");
```
`*error` instead of `error`

oni-link commented on src/nvim/event/process.h:
```diff
@@ -23,13 +23,15 @@ struct process {
   const char *cwd;
   char **argv;
-  Stream *in, *out, *err;
+  Stream in, out, err;
```
`process_init()` still initializes these members with `NULL`.

bfredl replied:
That explains gcc:s strange warning message (it complains of missing braces for all of Process, not the Stream members).

bfredl commented:
Rather than allowing a separate close callback `on_stdout: {"data": "Callback1", "close": "Callback2"}`maybe be should just indicate this with special value, like `data` set to `[]` in the same callback. But it might be a clumsy with all-data-on-close behavior. The best  It would also be good to distinguish signalling EOF in PTY input, maybe this can be `data=['']`. 

Another solution is to rename the `on_stdout`/`on_stderr` argument, so if the new name is used, another `flags` argument will be received, but the old name gives compatible behavior.

bfredl commented:
Implemented option `stdout_buffered` to get all output data at once (at EOF).

bfredl committed:
channels: generalize jobclose()

bfredl commented:
Now `jobclose(chan, part)` works for all channels. Obviously it should be renamed, just like `jobsend()`. Currently we have

Functions that return channels
====
`jobstart()`
`termopen()`
`sockconnect()`
`stdioopen()`

Functions that take jobs
===
`jobpid()`
`jobresize()`
`jobstop()`
`jobwait()`

Functions that take any channel
===
`rpcrequest()`
`rpcnotify()`
`jobsend()`
`jobclose()`

(ignoring deprecated functions that can be done using any above)

oni-link commented on src/nvim/eval.c:
```diff
@@ -5138,8 +5112,8 @@ bool garbage_collect(bool testing)
-    map_foreach_value(jobs, data, {
+    Channel *data;
+    map_foreach_value(channels, data, {
       set_ref_in_callback(&data->on_stdout, copyID, NULL, NULL);
```
Missing cast `(Callback *)` to hide `-Wincompatible_pointer_types` warning for first argument. 

Perhaps also add somewhere a `static_assert` to make sure that the `Callback` object is the first member in `CallbackReader`:
```c 
#include "nvim/assert.h"
#include <stddef.h>
STATIC_ASSERT(offsetof(CallbackReader,cb)== 0,"...");
```

bfredl replied:
I do get a warning already, I have just not fixed it as it is still WIP

oni-link commented on src/nvim/eval.c:
```diff
@@ -5138,8 +5112,8 @@ bool garbage_collect(bool testing)
+    Channel *data;
+    map_foreach_value(channels, data, {
       set_ref_in_callback(&data->on_stdout, copyID, NULL, NULL);
       set_ref_in_callback(&data->on_stderr, copyID, NULL, NULL);
```
Missing cast

oni-link commented on src/nvim/event/process.h:
```diff
@@ -43,9 +45,9 @@ static inline Process process_init(Loop *loop, ProcessType type, void *data)
-    .in = NULL,
-    .out = NULL,
-    .err = NULL,
+    .in = {0},
```
Clang complains here about a missing initializer (field `uv`). If we remove `.in = ...`, `.out =...` and `.err = ...` the compiler should set the fields to zero.

ZyX-I replied:
I prefer being explicit in such cases rather then rely on compiler zeroing things out.

bfredl replied:
Strange that clang complains, `{0}` is a standard pattern and it's used elsewhere without complaints.

bfredl commented:
What is also missing is able to write to nvim's own stderr. It should probably not be mixed into the stdio channel as `chansend(1, datas, "stderr")` would be quite ugly, and also stderr is always possible even if stdin/out is not used as a channel. Rather, either a dedicated channel `chansend(2, data)` or a specific function `write_stderr(data)`.

bfredl committed:
channels: stderr channel

oni-link commented on src/nvim/channel.c:
```diff
@@ -0,0 +1,499 @@
+}
+
+/// @param data will be consumed
+size_t channel_send(uint64_t id, char* data, size_t len, const char **error)
```
Why use `size_t` as return type? It looks like this function only returns `0` or `1`.

bfredl replied:
it should return the number of bytes written, as is standard for write function. But it looks like `wstream_write` doesn't do this.

oni-link replied:
`fwrite()` only returns the number of items written (here `1`). That number is also not equal the number of bytes written if `len != 1`.

oni-link commented on src/nvim/channel.c:
```diff
@@ -0,0 +1,499 @@
+  Channel *channel = channel_alloc(kChannelTypeSocket);
+  if (!socket_connect(&main_loop, &channel->stream.socket,
+                      tcp, address, timeout, error)) {
+    // TODO: early detroy that doesn't consume a channel id?
```
typo: detroy

bfredl replied:
A TODO without `(name)` is something I will fix before merging the PR.

oni-link commented on src/nvim/channel.c:
```diff
@@ -0,0 +1,499 @@
+  // TODO: test mixed mode (server RPC, client bytes)
+  if (!tcp && rpc) {
+    char *path = fix_fname(address);
+    if (rpc && server_owns_pipe_address(path)) {
```
`rpc` is here always true.

bfredl committed:
channels: update documentation

bfredl committed:
channels: tests

brcolow commented on runtime/doc/nvim_terminal_emulator.txt:
```diff
@@ -117,8 +117,8 @@ variables:
   running in the terminal can set this title via an escape sequence.
-- *b:terminal_job_id* The nvim job ID of the job running in the terminal. See
-  |job-control| for more information.
+- *b:terminal_job_id* The nvim channel ID or the PTY of the terminal emulateor. See
```
typo: emulateor -> emulator

oni-link replied:
Second half of the sentence not clear to me. What is `the PTY of the terminal emulator`?

bfredl replied:
should be `channel ID of the PTY of the emulatator`, but I guess it could be clearer; the channelrepresents specifically the master end of the PTY, not the PTY itself.

brcolow commented on runtime/doc/channels.txt:
```diff
@@ -0,0 +1,91 @@
+There are several ways to open a channel:
+
+  1. Through stdin/stdout when `nvim` is started with `--headless`, and a startup
+     script or --cmd  commad opens the stdio channel using |stdioopen()|.
```
typo: commad -> command

brcolow commented on runtime/doc/channels.txt:
```diff
@@ -0,0 +1,91 @@
+
+  4. By connecting to a TCP/IP socket or named pipe with |sockconnect()|.
+
+  5. By another process connecting to a socket listened by nvim. This only
```
socket listened by => socket listened to by

brcolow commented on runtime/doc/channels.txt:
```diff
@@ -0,0 +1,91 @@
+  5. By another process connecting to a socket listened by nvim. This only
+     supports RPC channels, see |rpc-connecting|.
+
+Channels can operate in two different modes, a mode using the |rpc| protocol,
```
I would re-organize to:

Channels can operate in two modes: one via the |rpc| protocol (msgpack-rpc
based) and the other in "bytes" mode where scripts read and write raw bytes
over the channel. As a caveat even if a job's channel is in |rpc| mode, bytes can still
be read over its' stderr.  Additionally only bytes can be written to nvim's stderr.

(Honestly it doesn't make a big difference either way, feel completely free to reject
this suggestion).

bfredl replied:
almost, but I prefer to think of stderr as an asset, not as a "caveat"

bfredl commented:
Maybe this would also be a good time to add `nvim_terminal_write_raw` , unless we think it's ok that people can and will abuse stderr to output custom sequences to the terminal.

oni-link commented on src/nvim/event/rstream.c:
```diff
@@ -105,19 +105,19 @@ static void read_cb(uv_stream_t *uvstream, ssize_t cnt, const uv_buf_t *buf)
+    if (cnt == UV_ENOBUFS || cnt == 0) {
+      return;
+    } else if (cnt == UV_EOF && uvstream->type == UV_TTY) {
+      // The TTY driver might signal TTY without closing the stream
```
Why keep reading from the stream if we mark it `EOF` later on? Has this something to do with input `^D`?

bfredl replied:
Yes, it is especially for `uvstream->type == UV_TTY`, where `^D` leading to `UV_EOF` does not mean that the stream closed.

bfredl commented:
I wonder if is bug: multiline `--cmd` seems to be interrupted by `endfunction`, otherwise it works fine. Same in vim. If we fix that, `nvim_command` will already be `nvim_source`.

ZyX-I commented:
@bfredl If I remember correctly what is going on, these are interrupted by `:function`. Specifically when creating a function it always consumes at least current line and `--cmd` (as well as `:execute`) have only *one* line to consume: nobody bothers to split that on `\n`. Worse, you can’t fix it generally by just creating a `exarg_T.getline` for such kind of commands which will split on `\n`: due to this deficiency `python << EOF` does not work at all, while using `execute "python {multiline command}"` or `execute "normal! i{multiline text}\e"` does (which in turn does not work in files), so if you want to deal with `:execute {multiline string}` being different from the same `{multiline string}` sourced from a file then you will end breaking multiple scripts (some of the kind I wrote myself).

This may be addressed on per-command basis though, but only to an extend which will make situation slightly better: e.g. you can allow multiple functions in one `:execute` or working `execute "python << EOF\n…\nEOF"`, but without removing previous `:execute "python {multiline string}"` behaviour.

Probably the easiest thing to do is supporting `:execute {readfile-style list}` and making tests use that: here you don’t have backwards compatibility to keep up with, can define an easily comprehendable `exarg_T.getline` and maintain consistency. And you don’t actually even need much code for this: `execute([…])` already works in this fashion, though for API you would probably want define two separate functions (API function that accepts `ArrayOf(String)` as altering `nvim_command` signature is incompatible and something like `get_list_line`, but for `ArrayOf(String)` in order to not do useless conversion).

bfredl commented:
A separate function/form is probably better than a half-fix constrained by backwards compatibility. But at least there should be a warning/error for `"func ... \n ... endfunc \n morecode"`.

Maybe API functions (at least new ones) taking binary-compatible strings should be made to support both forms, both blob with inline \n and \000 (best for lua and remote plugins), and vimL-style line list (needed for non-null-safe like vimL)

ZyX-I commented:
@bfredl It looks like “no warning” is intentional given the note “must be on a line on its own” in `:h :endfunction`. Would not be surprised to find out somebody is actually writing `function Foo(…)…\n…\nendfunction Foo`, even though I do not see documentation actually stating this being valid.

As for blobs/lists: as long as you don’t need strings in containers (for which case there are msgpack-special-dicts, though yet need to enable them because currently conversion to API types happens like they do not exist) you can just autoconvert lists to binary String blobs since types are already known (though if special dicts are to be enabled better go explicit and allow only those and no automatic list_T to String conversions). This applies for VimL and other dynamically typed languages with such distasteful property (are there any BTW? I know only about POSIX sh, bash and fish so far, but for ones’ sanity sake one better not write any Neovim plugins in any of those).

bfredl commented:
>  “Must be on a line on its own” 

and yet `foo\nendfunction` works but not `endfunction\nbar`.Even from  just `echo 'aa'\necho 'bb'\n` working, people could easily conclude `\n` works just like in a script. Even if the docs are clarified, not even showing a warning for `endfunction\nsomecommand` would be spacebar-heating levels of backwards compatibility. (I would even make it an error, but that should be discussed with upstream) After all vim does throw syntax errors, rather than having the code silently misbehave and asking people to read though all of the docs to find the error they did.

bfredl commented:
> as long as you don’t need strings in containers (for which case there are msgpack-special-dicts, though yet need to enable them because currently conversion to API types happens like they do not exist) you can just autoconvert lists to binary String blobs since types are already known 

In fact it can work also in well-typed containers, but not in untyped containers and Object. Also I would allow the the API function implementation to receive a list of lines (but marked that it really represents a single blob, not a list of blobs.) but non-vimL callers to send a blob, that would be the best for `nvim_source`, which one from lua likely would call as `nvim_source([[ .... ]])`.

 I think special dicts should only be required when dealing with foreign type systems, I wouldn't expect them needed when calling API functions or RPC functions at least for RPC clients built with nvim in mind.

ZyX-I commented:
> and yet foo\nendfunction works but not endfunction\nbar.Even from just echo 'aa'\necho 'bb'\n working, people could easily conclude \n works just like in a script. Even if the docs are clarified, not even showing a warning for endfunction\nsomecommand would be spacebar-heating levels of backwards compatibility. (I would even make it an error, but that should be discussed with upstream) After all vim does throw syntax errors, rather than having the code silently misbehave and asking people to read though all of the docs to find the error they did.

This is the mechanics of how `\n` works in file (it separates lines) and how it works everywhere else (exactly like `|`). I am not saying that this is a good thing, but you can’t have a *clear* fix either.

Though from what I have seen in plugins `:endfunction` command is normally followed either by comments (like `" }}}`) or bang, don’t think that errorring out should one place something else* there would be a great deal for Neovim, though Bram will probably not accept such change.

\* Specifically it looks like something easy:

```Patch
diff --git a/src/nvim/eval.c b/src/nvim/eval.c
index baa61a26b..982ddc483 100644
--- a/src/nvim/eval.c
+++ b/src/nvim/eval.c
@@ -19842,8 +19842,16 @@ void ex_function(exarg_T *eap)
 
       /* Check for "endfunction". */
       if (checkforcmd(&p, "endfunction", 4) && nesting-- == 0) {
-        if (line_arg == NULL)
+        if (*p == '!') {
+          p++;
+        }
+        p += (char_u *)strspn((const char *)p, " \t\r\n");
+        if (*p != NUL && *p != '"') {
+          emsgf(_(e_trailing2), p);
+        }
+        if (line_arg == NULL) {
           xfree(theline);
+        }
         break;
       }
 
diff --git a/src/nvim/globals.h b/src/nvim/globals.h
index 840be928c..6d1bd1de1 100644
--- a/src/nvim/globals.h
+++ b/src/nvim/globals.h
@@ -1131,6 +1131,7 @@ EXTERN char_u e_longname[] INIT(= N_("E75: Name too long"));
 EXTERN char_u e_toomsbra[] INIT(= N_("E76: Too many ["));
 EXTERN char_u e_toomany[] INIT(= N_("E77: Too many file names"));
 EXTERN char_u e_trailing[] INIT(= N_("E488: Trailing characters"));
+EXTERN char_u e_trailing2[] INIT(= N_("E488: Trailing characters: %s"));
 EXTERN char_u e_umark[] INIT(= N_("E78: Unknown mark"));
 EXTERN char_u e_wildexpand[] INIT(= N_("E79: Cannot expand wildcards"));
 EXTERN char_u e_winheight[] INIT(= N_(
```

I have not tested that though.

---

As a side note: I found `strspn` function in C99. Looks like rewriting things like `skipwhite` with it may be an optimization.

justinmk commented on src/nvim/channel.c:
```diff
@@ -0,0 +1,512 @@
+// This is an open source non-commercial project. Dear PVS-Studio, please check
```
Should this live in `src/nvim/event/`?

bfredl replied:
No. channels are a higher-level abstraction, for vimL (and future lua) consumption, above the low-level libuv-wrapping `event/` code.

justinmk commented on src/nvim/channel.c:
```diff
@@ -0,0 +1,512 @@
+///
+/// @param id The channel id
+/// @return true if successful, false otherwise
+bool channel_close(uint64_t id, ChannelPart part, const char **error)
```
Now this module and `src/nvim/msgpack_rpc/channel.c` are sharing the `channel_` namespace. That seems unpleasant.

justinmk commented on runtime/doc/channels.txt:
```diff
@@ -0,0 +1,91 @@
+  2. Through stdin, stdout and stderr of a process spawned by |jobstart()|.
+
+  3. Through the PTY master end of a PTY opened with
+     `jobstart(..., {'pty': v:true})` or |termopen()|.
```
`'type': {pty,term,...}` instead of multiple boolean options.

bfredl replied:
I wouldn't mind it. But maybe in a later PR? After all, this PR does not introduce any new types.

justinmk commented on src/nvim/eval.c:
```diff
@@ -15509,6 +15451,35 @@ static void f_sort(typval_T *argvars, typval_T *rettv, FunPtr fptr)
 }
 
+/// "stdioopen()" function
+static void f_stdioopen(typval_T *argvars, typval_T *rettv, FunPtr fptr)
```
Can't we have e.g. a generic `chanopen()` function, instead of `sockconnect`, `stdioopen`, ...

bfredl replied:
Why? `sockconnect()` and `stdioopen()` does completely different things. Trying to connect to network or named pipe is not the same as "using these already open fds this way". The only thing they have in common is returning handle in the same namespace (as also `jobstart()` does).

justinmk replied:
It's not that different, it's opening a channel. We don't need numerous different narrowly-defined functions to open a channel. A `chanopen` function can work like `jobstart`, by taking options that control the behavior.

bfredl replied:
jobstart always starts a process. `sockconnect()` always connects to some socket, possibly over the network, at least some external process will be notified somewhere. `stdioopen()` has no immediately external effects, just sets the mode of operation of known fds.  `sockconnect()` and `stdioopen()` is as different from each other as to `jobstart()`. And yes in actual _behavior_, not merely "how unix names things".

justinmk commented on src/nvim/eval/typval.c:
```diff
@@ -374,7 +374,7 @@ void tv_list_append_dict(list_T *const list, dict_T *const dict)
 ///                  string or NULL “empty” string.
 void tv_list_append_string(list_T *const l, const char *const str,
-                           const ptrdiff_t len)
+                           const ssize_t len)
```
we were avoiding ssize_t with the hope of compiling on MSVC

bfredl replied:
can't we define it if it is not present?

bfredl commented:
@ZyX-I throwing an error for `endfunction \n code` doesn't seem unclear to me, though I don't mind your suggestion of throwing error for the most either. Only thing I could think of: script doing `function MyFunc() ... endfunction MyFunc`

justinmk commented:
Instead of more `'rpc': v:true`, I suggest a string field, `'type': 'rpc'`. That is what `jobstart()` will use in the future.

> Implemented option stdout_buffered to get all output data at once (at EOF).

What if EOF is never received? Is there a timeout? Is it controllable?

> Maybe this would also be a good time to add nvim_terminal_write_raw , unless we think it's ok that people can and will abuse stderr to output custom sequences to the terminal.

I think it's ok. But why wouldn't it be ok? Writing directly to the TUI is already a hack. Letting it be "possible" but still "hacky" feels right. If we do offer a API function it should be named `nvim_ui_write`; let the behavior be UI-defined.

bfredl commented:
EOF will be assumed on process exit, or you can force it using `jobclose(..., 'stdout')`.  

The problem I imagine is code being too lazy to check if nvim is a terminal or GUI. Also in the future we might support detached TUI, in case writing to stderr won't work.

ZyX-I commented:
It appears that `strspn` used for `skipwhite` or `skipdigits` is not an optimization (well, at least as long they are uninlined even in which case that is still questionable; though I used release build with gcc-5.4.0 and optimization which is doing inlining of strspn for skipwhite (not for skipdigits)). No such good results as with `vim_strchr`, except that multiline functions turn into a single line.

ZyX-I commented:
> @ZyX-I throwing an error for endfunction \n code doesn't seem unclear to me, though I don't mind your suggestion of throwing error for the most either. Only thing I could think of: script doing function MyFunc() ... endfunction MyFunc

I did not actually see this in a plenty of plugins I have in ~/.vam (most of which I do not use and they are thus not updated, not that I have all of them though).

bfredl commented:
> I suggest a string field, 'type': 'rpc'. 

I am not against the idea, but `type` is ambiguous; channels are typed along two axes: the transport type (job,pty,socket,stdio) and the protocol type (bytes,:terminal,rpc), to borrow e.g. twisted's terminology. We should use two unambiguous terms for channel_info.

oni-link commented on src/nvim/msgpack_rpc/channel.c:
```diff
@@ -777,14 +621,14 @@ static bool is_valid_rpc_response(msgpack_object *obj, Channel *channel)
-  return kv_size(channel->call_stack) && response_id
-    == kv_A(channel->call_stack, kv_size(channel->call_stack) - 1)->request_id;
+  return kv_size(channel->rpc.call_stack) && response_id
+    == kv_A(channel->rpc.call_stack, kv_size(channel->rpc.call_stack) - 1)->request_id;
```
To shorten the expression macro `kv_last` could be used.

bfredl replied:
Fixed, thanks.

oni-link commented on src/nvim/msgpack_rpc/channel.c:
```diff
@@ -777,14 +621,14 @@ static bool is_valid_rpc_response(msgpack_object *obj, Channel *channel)
-  ChannelCallFrame *frame = kv_A(channel->call_stack,
-                             kv_size(channel->call_stack) - 1);
+  ChannelCallFrame *frame = kv_A(channel->rpc.call_stack,
+                             kv_size(channel->rpc.call_stack) - 1);
```
`kv_last` instead of `kv_A`?

bfredl commented:
Is there a better way to check that a result is `eq()` one of two possible valid values, than this:

    if not pcall(function() eq(expected, received) end) then
      eq(alternatively_expected, received)
    end

ZyX-I commented:
@bfredl Except that `pcall()` signature is `pcall(func, arg1, …)` so you don’t need the `function()` wrapper I am not aware of anything. But you may always construct a new assertion type if you think it will be needed always (so far I only once created `assert.just_fail` in unit test helpers). The list of available ones is at the end of https://github.com/Olivine-Labs/luassert/blob/master/src/assertions.lua.

oni-link commented on src/nvim/channel.c:
```diff
@@ -0,0 +1,583 @@
+      }
+      if (!channel->stream.stderr.closed) {
+        channel->stream.stderr.closed = true;
+        // Don't close on exit, in case late error messeages
```
typo in `messeages`

oni-link commented on src/nvim/channel.c:
```diff
@@ -0,0 +1,583 @@
+      xfree(chan->stream.pty.term_name);
+    }
+    channel_destroy_early(chan);
+    *status_out = proc->status;
```
Missing `return NULL;`?

bfredl replied:
(fixed)

oni-link commented on src/nvim/event/rstream.c:
```diff
@@ -187,6 +188,7 @@ static void read_event(void **argv)
     size_t count = (uintptr_t)argv[1];
     bool eof = (uintptr_t)argv[2];
+    stream->did_eof = eof;
     stream->read_cb(stream, stream->buffer, count, stream->cb_data, eof);
```
Do we need the extra argument `eof` in the callback? This information is now part of stream.

bfredl replied:
it is not the same information. `eof` is part of the data relayed to the callback, it is not an "extra argument".  `stream->did_eof` is to avoid sending a superfluous EOF callback in the job cleanup code (which ensures all output callbacks are sent before the exit callback).

bfredl commented:
@justinmk This might be a good time to discuss naming of functions. Personally I would use `chan` without `_` (which follows nvim convention (no `_`) and is distinct enough from vim `ch_`) for functions taking any channel and still `job` for functions only taking jobs. So renamed will then be

`jobsend` -> `chansend`
`jobclose` -> `chanclose`
`rpcrequest` -> `chanrequest` (or `chancall`)
`rpcnotify` -> `channotify`

ZyX-I commented:
I think this is actually better to have one convention, we are merging Vim patches after all. I would rather discuss not naming, but where to put such new features: how about dumping all VimL job/channel/etc functions and making them API functions instead, preferably in a separate file and not in api/vim.c?

ZyX-I commented:
Though no, with the current state of the API it is not possible (partials), but may be possible if I (or somebody else) finish API Reference’s.

bfredl commented:
`nvim_chan_` API is probably the right solution anyway to expose them natively to lua , if we like we can start with just callback reference that only works in vimL and lua, not over rpc. But for this PR I think it is enough scope to just use vimL `chan`. After all only two functions have directly wrong names that "must" be renamed, so only two more soon-to-be-deprecated names, as we should let rpc- be if we plan to replace with API later anyway.

oni-link commented on src/nvim/channel.h:
```diff
@@ -0,0 +1,138 @@
+static inline Channel *find_channel(uint64_t id)
+{
+  Channel *data = pmap_get(uint64_t)(channels, id);
+  if (!data) {
```
The whole function body could be replaced with `return pmap_get(uint64_t)(channels, id);`

oni-link commented on runtime/doc/channels.txt:
```diff
@@ -0,0 +1,89 @@
+user over the terminal interface (TUI). However when invoked with
+`--headless`, the TUI is not started and stdin and stdout can be used as a
+channel. To open the stdio channel |stdioopen()| must be called during
+|startup|, as there later will be no way of invoking a command. Here is a
```
With `` `startup` `` (instead of `|startup|`) we can jump to the `startup` help.

oni-link commented on runtime/doc/eval.txt:
```diff
@@ -2127,13 +2136,11 @@ isdirectory({directory})	Number	|TRUE| if {directory} is a directory
-jobresize({job}, {width}, {height})
+jobpid({id})			Number	Returns pid of a job.
+jobresize({id}, {width}, {height})
 				Number	Resize {job}'s pseudo terminal window
```
No `{job}` argument in function mentioned.

oni-link commented on runtime/doc/eval.txt:
```diff
@@ -2750,6 +2759,25 @@ changenr()						*changenr()*
 		redo it is the number of the redone change.  After undo it is
 		one less than the number of the undone change.
 
+chansend({id}, {data})					{Nvim} *jobsend()*
```
`*jobsend()*` -> `*chansend()*`

oni-link commented on runtime/doc/eval.txt:
```diff
@@ -2750,6 +2759,25 @@ changenr()						*changenr()*
 
+chansend({id}, {data})					{Nvim} *jobsend()*
+		Send data to channel {id}. For a job, it writes it to the
+		stdin of the process. For the stdio channel |channel-stdio|,
```
Missing `s` in `|channels-stdio|`

oni-link commented on runtime/doc/eval.txt:
```diff
@@ -2750,6 +2759,25 @@ changenr()						*changenr()*
+		it writes to Nvim's stdout.
+		Returns the number of bytes written if the write succeeded,
+		0 otherwise.
+		See |channel-bytes| for more information.
```
Missing `s` in `|channels-bytes|`

oni-link commented on runtime/doc/eval.txt:
```diff
@@ -2750,6 +2759,25 @@ changenr()						*changenr()*
+		{data} is a list, the items will be separated by newlines and
+		any newlines in an item will be sent as a NUL. A final newline
+		can be sent by adding a final empty string. For example: >
+			:call jobsend(id, ["abc", "123\n456", ""])
```
`jobsend` -> `chansend`

oni-link commented on runtime/doc/eval.txt:
```diff
@@ -4870,27 +4901,17 @@ jobresize({job}, {width}, {height})			{Nvim} *jobresize()*
+		|channel|, and a channel ID is returned on success. Use
+		|chansend()| (or |rpcnotify()| and |rpcrequest()| if "rpc" option
+		was used) to send data to stdin and |chanclose()| to close stdio
+		streams without stopping the job explicitsy.
```
typo in `explicitsy`

oni-link commented on runtime/doc/eval.txt:
```diff
@@ -4927,14 +4948,12 @@ jobstart({cmd}[, {opts}])				{Nvim} *jobstart()*
-		    |rpcnotify()| and |rpcrequest()| if "rpc" option was used)
-		    and |jobstop()|
+		  - The channel ID on success
 		  - 0 on invalid arguments or if the job table is full
```
`job table is full` does not seem relevant anymore.

oni-link commented on runtime/doc/eval.txt:
```diff
@@ -6198,13 +6217,11 @@ rpcstart({prog}[, {argv}])				   {Nvim} *rpcstart()*
+		Deprecated. This function stopped a job with RPC channel, and
+		additionally closed RPC sockets. Don't use this function,
+		instead use |jobstop()| to stop any job, and
+		|chanclose|(id, "rpc") to close RPC communitation without
```
typo in `communitation`

oni-link commented on runtime/doc/eval.txt:
```diff
@@ -6198,13 +6217,11 @@ rpcstart({prog}[, {argv}])				   {Nvim} *rpcstart()*
+		additionally closed RPC sockets. Don't use this function,
+		instead use |jobstop()| to stop any job, and
+		|chanclose|(id, "rpc") to close RPC communitation without
+		stopping the job. Use |chanclose|(id) to close any socket.
```
Missing help entry for `chanclose()` as a jump target.

oni-link commented on runtime/doc/eval.txt:
```diff
@@ -6892,15 +6909,19 @@ sockconnect({mode}, {address}, {opts})			 *sockconnect()*
+		Returns a |channel| ID. Close the socket with |chanclose()|.
+		Use |chansend()| to send data over a bytes socket, and
+		|rpcrequest()| and |rpcnotify()| to communicate with a RPC
+		soceket.
```
typo

oni-link commented on runtime/doc/eval.txt:
```diff
@@ -7052,6 +7073,26 @@ sqrt({expr})						*sqrt()*
+		In a nvim launched with the |--headless| option, this opens
+		stdin and stdout as a |channel|. This function can only be
+		invoked once per instance. See |channels-stdio| for more
+		information and examples. See also |v:stderrchan|
```
Will there be a `v:stderrchan` or should this be `v:stderr`?

oni-link commented on runtime/doc/eval.txt:
```diff
@@ -7052,6 +7073,26 @@ sqrt({expr})						*sqrt()*
+		information and examples. See also |v:stderrchan|
+
+		Returns a |channel| ID. Close the stdio descriptors with |chanclose()|.
+		Use |chansend()| to send data stdout, and
```
missing word: data `to` stdout

oni-link commented on runtime/doc/eval.txt:
```diff
@@ -7052,6 +7073,26 @@ sqrt({expr})						*sqrt()*
+		Returns a |channel| ID. Close the stdio descriptors with |chanclose()|.
+		Use |chansend()| to send data stdout, and
+		|rpcrequest()| and |rpcnotify()| to communicate with a RPC
+		soceket.
```
typo

oni-link commented on runtime/doc/eval.txt:
```diff
@@ -7052,6 +7073,26 @@ sqrt({expr})						*sqrt()*
+		soceket.
+
+		{opts} is a dictionary with these keys:
+		  |on_stdio|: callback invoked when stdin is written to.
```
`sockconnect()` uses `on_data`, should this be used here too? Otherwise `on_stdin`?

oni-link commented on runtime/doc/eval.txt:
```diff
@@ -7052,6 +7073,26 @@ sqrt({expr})						*sqrt()*
+		  rpc      : If set, |msgpack-rpc| will be used to communicate
+			     over stdio
+		Returns:
+		  - The channel ID on success (this is always 1)
```
Should there be a `v:stdio` set to `1` if this function was called?

bfredl replied:
Why? the function already returns the value. And one-time callable function signals that the channel is "owned by" the one that calls it, stdout is not expected to be used randomly by random plugins (but a fixed number is good for the external embedder's point of view) . But OTOH `v:stderr` tells that the channel "belongs" to nvim itself, and could be used by more than one (but it should still be used sparingly).

oni-link commented on src/nvim/channel.c:
```diff
@@ -0,0 +1,598 @@
+      chan->stream.pty.width = pty_width;
+    }
+    if (pty_height > 0) {
+      chan->stream.pty.width = pty_height;
```
Wrong member: `width` -> `height`.

oni-link commented on src/nvim/eval.c:
```diff
@@ -11537,18 +11474,18 @@ static void f_jobresize(typval_T *argvars, typval_T *rettv, FunPtr fptr)
 
-  if (data->proc.uv.process.type != kProcessTypePty) {
-    EMSG(_(e_jobnotpty));
+  if (data->type != kChannelTypeProc
```
`find_job()` has already verified that `data->type == kChannelTypeProc`, so this condition could be removed.

oni-link commented on src/nvim/channel.c:
```diff
@@ -0,0 +1,598 @@
+      // allow double close, even though we can't say what parts was valid.
+      return true;
+    }
+    *error = (const char *)e_invchan;
```
This function is sometimes  called with `error == NULL`.

oni-link commented on src/nvim/channel.c:
```diff
@@ -0,0 +1,598 @@
+
+    case kChannelTypeStderr:
+      if (part != kChannelPartAll) {
+        return false;
```
Why is no error message set? Could be problematic for the calling function.

oni-link commented on src/nvim/channel.c:
```diff
@@ -0,0 +1,598 @@
+    return 0;
+  }
+
+  if (did_stdio) {
```
`did_stdio` is always `false`. Is setting this variable on the TODO list?

oni-link commented on src/nvim/channel.c:
```diff
@@ -0,0 +1,598 @@
+uint64_t channel_from_stdio(bool rpc, CallbackReader on_output, const char **error)
+{
+  if (!headless_mode) {
+    *error = _("Stdio channel can only be opened in headless mode");
```
`error` could be `NULL`. Perhaps document, when it is safe to call this function with `error == NULL`?

oni-link commented on src/nvim/eval.c:
```diff
@@ -13941,10 +13857,14 @@ static void f_rpcstop(typval_T *argvars, typval_T *rettv, FunPtr fptr)
   } else {
-    rettv->vval.v_number = channel_close(argvars[0].vval.v_number);
+    const char *error;
+    if (!channel_close(argvars[0].vval.v_number, kChannelPartRpc, &error)) {
```
`rettv->vval.v_number` is not set anymore on success.

justinmk committed:
tui: job-control: use saved termios for pty jobs

On startup, if running in a terminal, save the termios properties.
Use the saved termios for `:terminal` and `jobstart()` pty jobs.

This won't affect nvim spawned outside of a terminal.

questions:

- This affects `:terminal` and `jobstart({'pty':v:true})`.
  Should we be more conservative for `jobstart({'pty':v:true})` (e.g.
  pass NULL to forkpty() and let the OS defaults prevail)?
  - Note: `iutf8` would not be set in that case.

bfredl committed:
channels: move away term code from eval.c

bfredl commented:
@justinmk IIRC you wanted to get rid of the `jobstart` / `termopen` distiction. One way of 
 doing that could be to decouple the lifetime of the channel from the terminal. In the new API we will instead have `chan = nvim_job_start(..., {'pty':true}); buf = nvim_chan_terminal_open(chan)`. With the guarantee if the calls are chained no output will be "lost" before terminal capture begins. Also I never understood the awkward API of `termopen()` implicitly consuming a pristine, unmodified curbuf and converting it into a terminal. Instead it should just allocate a terminal buffer and return it, and the API consumer can make it current or whatever.

justinmk commented:
`f_termopen` as currently implemented is mostly an alias for `f_jobstart`. I already eliminated `termopen()` in https://github.com/neovim/neovim/pull/5485 (though unfortunately I didn't wrap up that PR yet) by adding a `type={pty,term}` option to `jobstart()`.

> In the new API we will instead have chan = nvim_job_start(..., {'pty':true}); buf = nvim_chan_terminal_open(chan). 

Why not `nvim_job_start(..., {'type':'term'})`. Why is the 2nd step needed? Why is a terminal-specific family of functions needed? 

> With the guarantee if the calls are chained no output will be "lost" before terminal capture begins. 

By "chained" do you mean adjacent?

> Also I never understood the awkward API of termopen() implicitly consuming a pristine, unmodified curbuf and converting it into a terminal. Instead it should just allocate a terminal buffer and return it, and the API consumer can make it current or whatever.

Agree, I was thinking that assigning a buffer as the stdout "sink" and/or stdin "source" for any job would be a way to generalize it.

justinmk commented on runtime/doc/nvim_terminal_emulator.txt:
```diff
@@ -117,8 +117,8 @@ variables:
-- *b:terminal_job_id* The nvim job ID of the job running in the terminal. See
-  |job-control| for more information.
+- *b:terminal_channel_id* The nvim channel ID for the underlying PTY.
+  |chansend()| can be used to send input to the terminal.
```
If we're going to introduce a new variable it should be `b:term_` prefix like `b:term_title`.

But I'm not sure these need to be terminal-specific: why not have buffer-local `'jobid'` or `'chanid'` options which control what job drives a buffer?

bfredl replied:
`chanid` option is probably better than polluting `b:` further. `chanid` option is more general than `jobid` as we in theory could allow a socket or stdin to be directly connected to a buffer.

bfredl commented:
> Why is the 2nd step needed? Why is a terminal-specific family of functions needed?

Two steps are what is required today: `noautocmd enew | call termopen(...)`

A function is not a family, of course it will belong to the `nvim_chan_` family. As I already said, it enables to later on connect a `:terminal` to an existing pty channel. The alternative would be to have `nvim_job_start` to dynamically change its output to two outputs (a channel and a buffer), which I think is more awkward. For sure, an API function should never implicitly take over curbuf.

Also we should never use `'type'`, channels are typed on two axes, so we should rather have `'transport'` and `'protocol'` or something like that. 

> By "chained" do you mean adjacent?

indeed.

justinmk commented:
>  As I already said, it enables to later on connect a :terminal to an existing pty channel.

But why only terminal jobs? 

> Also we should never use 'type', channels are typed on two axes, so we should rather have 'transport' and 'protocol' or something like that.

Too pedantic. "type" is easier for users to understand. No one will be confused if other fields also change characteristics of the job/channel.

bfredl commented:
> But why only terminal jobs?

I don't follow, only terminals are terminals. If you mean non-pty jobs connected to buffers somehow, it shouln't be lumped togheter with `:termilal` (also I'm not sure if we want that, or rather "whatever" vimL functionally needed to emulate vim 8's various channel-buffer behaviors, which then can be reused for other purposes.


Dealing with reality is not "too pedantic"

> "type" is easier for users to understand. 

What user? How do you know most users consider would consider "type" to mean "transport"? Or that most users would consider it to mean "protocol"? You haven't even stated which...

> No one will be confused if other fields also change characteristics of the job/channel.

This makes my point, no single field should arbitrarily be considered the most important "type" field. Rather a field should simply be named after what it actually does/describes. (BTW "transport" and "protocol" can surely be improved, but it must be clear which is which).

brcolow commented on runtime/doc/channel.txt:
```diff
@@ -56,6 +56,8 @@ only the complete output is useful, and not partial data.
   0: The channel id
   1: the raw data read from the channel, formatted as a |readfile()|-style
      list. If EOF occured, a single empty string `['']` will be passed in.
+     Note that the items in this list doesn't (necessarily) correspond to
```
doesn't => don't

oni-link commented on src/nvim/eval.c:
```diff
@@ -16682,16 +16631,16 @@ static void f_termopen(typval_T *argvars, typval_T *rettv, FunPtr fptr)
   (void)setfname(curbuf, (char_u *)buf, NULL, true);
```
In the old code `terminal_open()` was only called after the term title was created and saved.
Now `channel_terminal_open()` (further up) calls `terminal_open()` before the term title is ready.

oni-link commented on src/nvim/eval.c:
```diff
@@ -16682,16 +16631,16 @@ static void f_termopen(typval_T *argvars, typval_T *rettv, FunPtr fptr)
   // Save the job id and pid in b:terminal_job_{id,pid}
   Error err = ERROR_INIT;
+  dict_set_var(curbuf->b_vars, cstr_as_string("terminal_channel_id"),
+               INTEGER_OBJ(chan->id), false, false, &err);
```
Missing call to `api_clear_error(&err);`. If an error is set here and in the next `dict_set_var()`, we have a memory leak.

oni-link commented on src/nvim/eval.lua:
```diff
@@ -55,6 +55,8 @@ return {
     ceil={args=1, func="float_op_wrapper", data="&ceil"},
     changenr={},
+    chanclose={args={1, 2}},
+    chansend={args={1, 2}},
```
`chansend()` needs always two arguments.

oni-link commented on src/nvim/eval/typval.c:
```diff
@@ -847,6 +847,30 @@ bool tv_callback_equal(const Callback *const cb1, const Callback *const cb2)
+    case kCallbackNone: {
+      break;
+    }
+    default: {
```
Remove default case?

brcolow commented on runtime/doc/channel.txt:
```diff
@@ -0,0 +1,141 @@
+read over its stderr. Similarily, only bytes can be written to nvim's own stderr. 
+
+A callback function will be invoked with data read from the channel.
+By default, the callback will invoked immediately when data is available, to
```
"the callback will invoked" => "the callback will be invoked"

brcolow commented on runtime/doc/channel.txt:
```diff
@@ -0,0 +1,141 @@
+  0: The channel id
+  1: the raw data read from the channel, formatted as a |readfile()|-style
+     list. If EOF occured, a single empty string `['']` will be passed in.
+     Note that the items in this list doesn't (necessarily) correspond to
```
"doesn't" => "don't"

brcolow commented on runtime/doc/channel.txt:
```diff
@@ -0,0 +1,141 @@
+     list. If EOF occured, a single empty string `['']` will be passed in.
+     Note that the items in this list doesn't (necessarily) correspond to
+     actual lines in the output. See |channel-lines|
+  2: The event type. This is to allow multible on_{event} handlers to
```
"multible" => "multiple"

brcolow commented on runtime/doc/channel.txt:
```diff
@@ -0,0 +1,141 @@
+     Note that the items in this list doesn't (necessarily) correspond to
+     actual lines in the output. See |channel-lines|
+  2: The event type. This is to allow multible on_{event} handlers to
+     implemented by the same function. The available events depends on how the
```
"to implemented" => "to be implemented"

brcolow replied:
"depends" => "depend"

brcolow commented on runtime/doc/channel.txt:
```diff
@@ -0,0 +1,141 @@
+used as callbacks.
+
+Data can be sent to the channel using the |chansend()| function. Here is a
+simple example echoing some data through a cat-process:
```
"a simple example echoing" => "a simple example of echoing"

brcolow commented on runtime/doc/channel.txt:
```diff
@@ -0,0 +1,141 @@
+    call chansend(id, "hello!")
+
+
+Here is a example setting a buffer to the result of grep, but only after all
```
"Here is a example setting" => "Here is an example of setting"

brcolow commented on runtime/doc/channel.txt:
```diff
@@ -0,0 +1,141 @@
+
+TODO
+
+For additional examples with jobs, see |job-control|.
```
unnecessary comma after "jobs"

brcolow commented on runtime/doc/channel.txt:
```diff
@@ -0,0 +1,141 @@
+For additional examples with jobs, see |job-control|.
+
+							      *channel-pty*
+As a special case is PTY channels opened by  `jobstart(..., {'pty': v:true})` .
```
"As a special case is" => "As a special case"

brcolow commented on runtime/doc/channel.txt:
```diff
@@ -0,0 +1,141 @@
+
+When channels are opened with the `rpc` option set to true, the channel can be
+used for remote method calls in both directions, see |msgpack-rpc|. Note that
+rpc channels are implicitly trusted and the process at the other end caninvoke
```
"caninvoke" => "can invoke"

brcolow commented on runtime/doc/channel.txt:
```diff
@@ -0,0 +1,141 @@
+As a convenience, the stdio channel will always have channel id 1.
+
+Nvim's stderr is implicitly open as a write-only, bytes channel. It will
+always have channel id 2, however to be explicit|v:stderr|can be used.
```
Should "|v:stderr|" have spaces around it?

brcolow commented on runtime/doc/eval.txt:
```diff
@@ -2751,6 +2760,35 @@ changenr()						*changenr()*
+			:call chansend(id, ["abc", "123\n456", ""])
+< 		will send "abc<NL>123<NUL>456<NL>".
+
+		If channel was opened with the rpc option this function
```
"If channel" => "If a channel"

brcolow commented on runtime/doc/eval.txt:
```diff
@@ -6199,13 +6218,11 @@ rpcstart({prog}[, {argv}])				   {Nvim} *rpcstart()*
-		killing the job.
-		Closes the socket connection if the channel was opened by
-		connecting to |v:servername|.
+		Deprecated. This function stopped a job with RPC channel, and
```
Very minor issue, but does it make sense to use the past tense for describing what a deprecated function did? Does it still do it?

bfredl replied:
maybe "was used to stop"

brcolow commented on runtime/doc/msgpack_rpc.txt:
```diff
@@ -61,24 +60,24 @@ To get a formatted dump of the API using python (requires the `pyyaml` and
-     process's stdin and stdout. stderr can still be used, though.
+An rpc socket is automatically created with each instance. The socket
+ location is stored in |v:servername|. By default this is a named pipe
+with an autmatically generated adress. See |XXX|.
```
"adress" => "address"

brcolow commented on runtime/doc/msgpack_rpc.txt:
```diff
@@ -61,24 +60,24 @@ To get a formatted dump of the API using python (requires the `pyyaml` and
+   |$NVIM_LISTEN_ADDRESS| environment variable before starting Nvim: >
         NVIM_LISTEN_ADDRESS=127.0.0.1:6666 nvim
-<
+<Also, more sockets and named pipes can be listened using |serverstart()|.
```
"listened" => "listened on" or "listened to" ?

brcolow commented on runtime/doc/msgpack_rpc.txt:
```diff
@@ -61,24 +60,24 @@ To get a formatted dump of the API using python (requires the `pyyaml` and
-<
+<Also, more sockets and named pipes can be listened using |serverstart()|.
+
+Note that localhost TCP sockets generally are less secure than named pipes,
```
"generally are" => "are generally"

ZyX-I commented on runtime/doc/eval.txt:
```diff
@@ -6200,13 +6219,11 @@ rpcstart({prog}[, {argv}])				   {Nvim} *rpcstart()*
-		killing the job.
-		Closes the socket connection if the channel was opened by
-		connecting to |v:servername|.
+		Deprecated. This function was used to stop a job with |rpc|
```
Code was broken so it now no longer can be used to stop such job? No past tense unless you want to document deprecated *and* removed functionality because with past tense it looks exactly like this, no matter just past tense or with “was used to”.

bfredl replied:
It is backwards compatible so far that old code using rpcstart/rpcstop or calling rpcstop on remotely connected channels still should work as expected, but its semantics is very confusing together with the new channels design: it does something different if the channel happened to be a job or not. So the wording is deliberately somewhat stronger than eg `rpcstart`, which while inconsistent api, will not do anything unexpected when used together with the newer functions.

justinmk replied:
should be moved to deprecated.txt

oni-link commented on src/nvim/msgpack_rpc/channel.c:
```diff
@@ -776,15 +587,16 @@ static bool is_rpc_response(msgpack_object *obj)
   // Must be equal to the frame at the stack's bottom
-  return kv_size(channel->call_stack) && response_id
-    == kv_A(channel->call_stack, kv_size(channel->call_stack) - 1)->request_id;
+  return kv_size(channel->rpc.call_stack)
```
If the stack size can be zero then `kv_last()` should not be used for `frame` (negative array index, UB?).

bfredl committed:
Revert channel logging, rebased on new code below

bfredl committed:
channels: reimplement logging (as stub for proper event)

bfredl committed:
channels: refactor jobwait

oni-link commented on src/nvim/eval.c:
```diff
@@ -11745,28 +11665,31 @@ static void f_jobwait(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+
   list_T *args = argvars[0].vval.v_list;
-  list_T *rv = tv_list_alloc();
+  Channel **jobs = xcalloc(sizeof(*jobs), args->lv_len);
```
For the result it does not matter, but the arguments of `xcalloc()` are switched.

bfredl replied:
Missing a free also..

oni-link commented on src/nvim/rbuffer.c:
```diff
@@ -121,7 +121,7 @@ char *rbuffer_read_ptr(RBuffer *buf, size_t *read_count) FUNC_ATTR_NONNULL_ALL
   if (!buf->size) {
     *read_count = 0;
-    return NULL;
+    return buf->read_ptr;
```
Merged PR #7089 introduced some `NULL` checks that could be removed because of this change.

bfredl replied:
Wouldn't they still need to check nonzero size? Otherwise will they not deref an initialized buffer instead in the situation it crashed?

oni-link replied:
Yes, nonzero size still needs to be checked, but some(/all) of the `NULL` checks could be removed. `out_data_append_to_screen()`  could also be changed, because this function is now never called with a `NULL` pointer for `output` (`FUNC_ATTR_NONNULL_ALL`?).

oni-link commented on src/nvim/channel.h:
```diff
@@ -77,7 +77,8 @@ struct Channel {
 
-  varnumber_T *status_ptr; // TODO: refactor?
+  // Exit status. Note: may be overwritten by f_jobwait.
+  varnumber_T status;
```
Is this member needed in the channel struct? Only channels that use `proc` have a status and that status could be directly read from `channel->stream.proc->status`. It looks like that this variable is only here to simplify `f_jobwait()` a bit.

bfredl replied:
Good point. We could allow it to be overwritten -2 as well.

oni-link commented on src/nvim/channel.c:
```diff
@@ -0,0 +1,740 @@
+      shell_free_argv(argv);
+      channel_destroy_early(chan);
+      *status_out = 0;
+      return NULL;
```
`pty_name` is not freed. Triggered by
`:echo jobstart('ls',{'pty':1,'detach':1,'TERM':"lsdjfslkdfjslkdfjsldkfklsdfsd"})`

oni-link commented on src/nvim/os/shell.c:
```diff
@@ -207,16 +207,15 @@ static int do_os_system(char **argv,
+  int status = process_spawn(proc, input != NULL, true, true);
+  proc->in.events = events;
+  proc->out.events = events;
+  proc->err.events = events;
```
Changing `proc->{in,out,err}.events` from `NULL` to `events` should not be necessary.

oni-link commented on src/nvim/os/shell.c:
```diff
@@ -236,27 +235,27 @@ static int do_os_system(char **argv,
+  proc->out.events = NULL;
+  rstream_init(&proc->out, 0);
+  rstream_start(&proc->out, data_cb, &buf);
+  proc->err.events = NULL;
```
`proc->{in,out,err}.events = NULL` could be removed, if previous set was removed.

bfredl replied:
You right, strange it was written this in way to start with. (it made even less before when the streams weren't event owned by the proc)

oni-link commented on src/nvim/channel.h:
```diff
@@ -0,0 +1,135 @@
+#define CHAN_STDERR 2
+
+#define CHAN_STDIO 1
+#define CHAN_STDERR 2
```
Both defines are duplicates.

oni-link commented on src/nvim/eval.c:
```diff
@@ -15151,19 +15067,19 @@ static void f_sockconnect(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+    if (!tv_dict_get_callback(opts, S_LEN("on_data"), &on_data.cb)) {
+      return;
+    }
+    on_data.buffered = tv_dict_get_number(opts, "data_buffered");
```
`data_buffered` is not mentioned in the help for `sockconnect()`

oni-link commented on src/nvim/eval.c:
```diff
@@ -22390,318 +22313,48 @@ char_u *do_string_sub(char_u *str, char_u *pat, char_u *sub,
+      &&tv_dict_get_callback(vopts, S_LEN("on_stderr"), &on_stderr->cb)
       && tv_dict_get_callback(vopts, S_LEN("on_exit"), on_exit)) {
+    on_stdout->buffered = tv_dict_get_number(vopts, "stdout_buffered");
+    on_stderr->buffered = tv_dict_get_number(vopts, "stderr_buffered");
```
`stdout_buffered` and `stderr_buffered` are not mentioned in the help for `jobstart()/termopen()`.

oni-link replied:
We could be buffering data without ever using it:
```
:echo termopen('bash',{'stdout_buffered':1})
```
Output for a pty would be buffered, but no callback was specified.

justinmk commented on src/nvim/event/stream.c:
```diff
@@ -90,6 +90,13 @@ void stream_close(Stream *stream, stream_close_cb on_stream_close, void *data)
   }
 }
 
+void stream_may_close(Stream *stream)
```
"try" would be more appropriate: `stream_try_close`

("maybe" usually describes the state of a thing, rather than the intent of a function)

justinmk commented on src/nvim/msgpack_rpc/channel.c:
```diff
@@ -768,14 +761,6 @@ static Channel *register_channel(ChannelType type, uint64_t id,
-        : (type == kChannelTypeSocket ? "socket"
-           : (type == kChannelTypeStdio ? "stdio"
-              : (type == kChannelTypeInternal ? "internal" : "?")))),
-       (source ? source : "?"));
```
why revert the logging?

bfredl commented:
As the commit says, it is reimplemented later. Less work than dealing with
all the rebase conflicts

Den 18 aug. 2017 10:00 em skrev "Justin M. Keyes" <notifications@github.com
>:

> *@justinmk* commented on this pull request.
> ------------------------------
>
> In src/nvim/msgpack_rpc/channel.c
> <https://github.com/neovim/neovim/pull/6844#discussion_r134043769>:
>
> > @@ -768,14 +761,6 @@ static Channel *register_channel(ChannelType type, uint64_t id,
>    kv_init(rv->call_stack);
>    kv_init(rv->delayed_notifications);
>    pmap_put(uint64_t)(channels, rv->id, rv);
> -
> -  ILOG("new channel %" PRIu64 " (%s): %s", rv->id,
> -       (type == kChannelTypeProc ? "proc"
> -        : (type == kChannelTypeSocket ? "socket"
> -           : (type == kChannelTypeStdio ? "stdio"
> -              : (type == kChannelTypeInternal ? "internal" : "?")))),
> -       (source ? source : "?"));
>
> why revert the logging?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/neovim/neovim/pull/6844#pullrequestreview-57298809>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABTMoCto5ZQp9hny3ifku2z8hGEsRyK0ks5sZe1lgaJpZM4NvXVM>
> .
>

justinmk commented on src/nvim/event/stream.c:
```diff
@@ -100,6 +102,8 @@ void stream_may_close(Stream *stream)
   FUNC_ATTR_NONNULL_ALL
 {
+  assert(!stream->uv_closed);
+  stream->uv_closed = true;
```
What's going on here? We are tracking two different closed states?

bfredl replied:
Deadlock on exiting is what is going on. `closed` appearently doesn't mean "closed" always, only "intent on close" as soon as the internal refcount goes to zero. But in mch_exit we cannot wait, we must close everything _now_ (and we won't return to the event handler holding the refcount anyway).

Alternatively we could refactor exiting to not invoke arbitrary recursive event handling (a more "do what's really necessary and then exit(3)" style of exiting) but that's probably another PR...

bfredl replied:
The situation occurs because "press RETURN". We "formally" forbid event processing at press-RETURN, so it "safely" can happen even in any event handler, but that's very much a false promise, as nvim could always receive an exiting signal at press-RETURN, which as mentioned, invokes arbitrary event processing...

bfredl committed:
channels: refactor

bfredl commented:
@justinmk exactly why does `receive_msgpack` check if the _out_-stream were closed, and not close_cb for the out stream where it belongs? We will never get sane event handling if random callback handlers check for random unrelated events. It already did check for instream closing via the `eof` argument, so that added check was completely redundant.

bfredl committed:
channels: more consistent event handling

terminal: libvterm now receives data in async context. This was "almost" safe
already, as redraws were queued anyway.

oni-link commented on runtime/doc/eval.txt:
```diff
@@ -4964,7 +4985,9 @@ jobstart({cmd}[, {opts}])				{Nvim} *jobstart()*
 		  |on_stdout|: stdout event handler (function name or |Funcref|)
+		  stdout_buffered : read stdout in |buffered| mode.
 		  |on_stderr|: stderr event handler (function name or |Funcref|)
+		  stderr_buffered : read stdout in |buffered| mode.
```
read stderr ...

justinmk commented:
@bfredl Didn't I explain it in https://github.com/neovim/neovim/pull/7081#issuecomment-323518777 ? Or you're talking about something else?

bfredl commented:
Even if stream events are "incorrectly queued" they should still be correctly ordered for streams belonging to the same queue.

Checks against impossible conditions does make the code harder to read. They should be marked with `assert()` or `abort()` at least.

> Doesn't that assume rstream.c:read_cb() was queued and processed first? IIRC that wasn't always the case.

What you do mean? Your added check checked if the stream was closed, but if the check is true you _literally_ are in the close callback _right_ _now_.

justinmk commented:
I've explained what I observed. Additionally, I wasn't comfortable adding a check only for one type of stream, and the condition there serves as some form of documentation, like an assertion. I expect you will disagree with that as well. If the tests pass then it doesn't matter much.

bfredl commented:
>  If the tests pass then it doesn't matter much

It doesn't matter for the current state of things, but it does matter when people try to understand the code to change/improve it. At least we should not have check for the same condition twice in a row, differently formulated, because then people will interpret it as "hmm are these two maybe different conditions?"

justinmk commented:
Actually, I thought the entire check in `receive_msgpack` would be able to be removed in this PR.

bfredl commented:
Against the particular condition in the test in #7081 yes (is fixed by more direct event handling), but I wonder what happens if RPC job randomly closes its stdin (without closing stdout) and then the channel gets written by an _unrelated_ event, that might need a check in `channel_write` or `close_cb` of the instream. Shouldn't be too hard to test.

bfredl committed:
test: add hexdump utilitiy function

bfredl committed:
core dumps: don't use pipe, it does not work

bfredl committed:
input: only change mode of input fd if there is an input fd

bfredl commented:
almost all tests pass now, clint looks like false positive. Marking RFC. still some missing docs work.

oni-link commented:
When doing the following,
```
nvim -u NONE -i NONE
"+yy
:qa
```
 I see the error message 'clipboard: error invoking xsel: ' on exit.

bfredl commented:
The underlying bug is present on master also: exit_cb is incorrectly called when exiting even for detached processes. However this branch calls it -1 which while more truthful the vimscript interprets an error exit...

bfredl commented:
should be fixed now

justinmk commented on src/nvim/channel.c:
```diff
@@ -639,7 +639,11 @@ static void channel_process_exit_cb(Process *proc, int status, void *data)
 
-  process_job_event(chan, &chan->on_exit, "exit", NULL, 0, status);
+  // if status is -1 the process did not really exit,
+  // we just closed the handle onto a detached process
```
What if the process return -1?

bfredl replied:
Doesn't POSIX only guarantee 0-255 works? If I try `int main() { return -1; }` I get 255 back.

oni-link commented on src/nvim/if_cscope.c:
```diff
@@ -778,7 +778,6 @@ static int cs_create_connection(size_t i)
     if (execl("/bin/sh", "sh", "-c", cmd, (char *)NULL) == -1)
       PERROR(_("cs_create_connection exec failed"));
 
-    stream_set_blocking(input_global_fd(), true);  // normalize stream (#2598)
```
Why could this be removed?

bfredl replied:
Because it made no sense in the first place? We are in a subprocess with replaced fds, so `input_global_fd()` doesn't mean anything.

oni-link commented on src/nvim/os/pty_process_unix.c:
```diff
@@ -36,23 +36,37 @@
+///
+/// @param tty_fd   TTY file descriptor, or -1 if not in a terminal.
+void pty_process_save_termios(int tty_fd)
+  FUNC_ATTR_NONNULL_ALL
```
Could be removed, no pointer arguments.

oni-link commented on src/nvim/msgpack_rpc/channel_defs.h:
```diff
@@ -0,0 +1,39 @@
+  uint64_t next_request_id;
+  kvec_t(ChannelCallFrame *) call_stack;
+  kvec_t(WBuffer *) delayed_notifications;
+  MultiQueue *events;
```
This member is never used.

oni-link commented on src/nvim/msgpack_rpc/channel.c:
```diff
@@ -304,112 +202,52 @@ void channel_subscribe(uint64_t id, char *event)
-  }
-
   size_t count = rbuffer_size(rbuf);
   DLOG("ch %" PRIu64 ": parsing %u bytes from msgpack Stream: %p",
```
`count` is of type `size_t` so we have to use `%zu` instead of `%u`.

justinmk commented on test/helpers.lua:
```diff
@@ -225,21 +225,7 @@ local function check_cores(app, force)
-      out:write('Failed to create pipe\n')
-    end
-    out:flush()
+    os.execute(db_cmd:gsub('%$_NVIM_TEST_APP', app):gsub('%$_NVIM_TEST_CORE', core) .. ' 2>&1')
```
Looks like now the backtrace is not written to stdout? Or does `os.execute()` do that implicitly?

bfredl replied:
It implicitly does nothing, which means the subprocess inherits stdio.

ZyX-I replied:
You are missing `out:write('\n')`, AFAIR it was there for a reason, though do not remember whether I just wanted an empty line or the above command sometimes did not produce trailing newline.

And code looks like I was planning to write output to some configurable location instead of simply dumping to stdout always, do not remember what for I needed that exactly.

bfredl replied:
added `\n`. We can revert and debug the pipe code once there is an actual need for it.

oni-link commented on src/nvim/msgpack_rpc/channel.c:
```diff
@@ -731,101 +535,44 @@ static void unsubscribe(Channel *channel, char *event)
-/// Close the channel streams/process and free the channel resources.
-static void close_channel(Channel *channel)
+
+/// Release the rpc infastructures reference to the channel
```
typo: infastructures -> infrastructure
`rpc` is released(freed?) with the channel, perhaps change comment to something like  `Mark rpc as closed, prepare to free channel.`

oni-link commented on src/nvim/msgpack_rpc/channel.c:
```diff
@@ -731,101 +535,44 @@ static void unsubscribe(Channel *channel, char *event)
-static void close_channel(Channel *channel)
+
+/// Release the rpc infastructures reference to the channel
+/// Don't call this directly, call channel_close(id, kChannelPartRpc, &error)
```
Is this comment still relevant? `channel_process_exit_cb` calls `rpc_close()` directly.

bfredl replied:
The other way around, that call can be removed, as `receive_msgpack` should've been called with `eof` before `exit_cb`

oni-link commented on src/nvim/msgpack_rpc/channel.c:
```diff
@@ -731,101 +535,44 @@ static void unsubscribe(Channel *channel, char *event)
-      // nothing to free.
-      break;
+  if (channel->streamtype == kChannelStreamStdio) {
+    multiqueue_put(main_loop.fast_events, exit_event, 1, channel);
```
`exit_event()` does not use `channel` anymore, so we could use `multiqueue_put(main_loop.fast_events, exit_event, 0)`.

bfredl commented:
> We could be buffering data without ever using it

Yes. What I think we should do if `buffered` is specified but no stream`cb` is to put it in a field in the `self` dict so it is available to the exit callback, or say code `jobwait()`-ing it.

oni-link commented on src/nvim/channel.c:
```diff
@@ -0,0 +1,736 @@
+    rstream_init(&proc->err, 0);
+    rstream_start(&proc->err, on_job_stderr, chan);
+  }
+  channel_create_event(chan, NULL);
```
This function checks `chan->term` for a mode description, but at this point the terminal is not yet opened. Only after calling `channel_terminal_open()` is `chan->term` set.

bfredl committed:
channels: improvements to buffering

oni-link commented on src/nvim/channel.c:
```diff
@@ -0,0 +1,751 @@
+  Channel *channel = channel_alloc(kChannelStreamStdio);
+
+  rstream_init_fd(&main_loop, &channel->stream.stdio.in, 0, 0);
+  wstream_init_fd(&main_loop, &channel->stream.stdio.out, 1, 0);
```
Replacing `0` and `1` with `OS_STDIN_FILENO` and `OS_STDOUT_FILENO`?

oni-link commented on src/nvim/channel.c:
```diff
@@ -0,0 +1,751 @@
+{
+  ChannelEvent *ev = (ChannelEvent *)args[0];
+
+  typval_T argv[4];
```
For what is the last field used?

bfredl replied:
For something somewhere in the `eval.c` code for invoking functions. Mentioned at `call_func`.

oni-link commented on src/nvim/channel.c:
```diff
@@ -0,0 +1,751 @@
+  typval_T argv[4];
+
+  argv[0].v_type = VAR_NUMBER;
+  argv[0].v_lock = 0;
```
For all fields: `argv[...].v_lock = VAR_UNLOCKED;`

bfredl replied:
The old code in `eval.c` had `v_lock = 0`. What has changed since then?

oni-link replied:
Since 18903bd9b8 `.v_lock` has an enum type.

oni-link commented on src/nvim/eval.c:
```diff
@@ -11695,32 +11632,19 @@ static void f_jobstart(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+  Channel *chan = channel_job_start(argv, on_stdout, on_stderr, on_exit, pty,
+                                    rpc, detach, cwd, width, height, term_name,
+                                    &rettv->vval.v_number);
+  channel_create_event(chan, NULL);
```
If job start fails, this function cannot be called:
```
:echo jobstart('bash',{'pty':1,'detach':1})
```

oni-link commented on src/nvim/eval.c:
```diff
@@ -13952,10 +13865,11 @@ static void f_rpcstart(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+                                    CALLBACK_READER_INIT, CALLBACK_NONE,
+                                    false, true, false, NULL, 0, 0, NULL,
+                                    &rettv->vval.v_number);
+  channel_create_event(chan, NULL);
```
Function cannot be called if job start failed.

oni-link commented on src/nvim/channel.c:
```diff
@@ -0,0 +1,751 @@
+        ga_clear(&reader->buffer);
+      } else if (reader->self) {
+        list_T *data = buffer_to_tv_list(reader->buffer.ga_data,
+                                         (size_t)reader->buffer.ga_len);
```
`buffer_to_tv_list()` has attribute `FUNC_ATTR_NONNULL_ALL`, but if no output was seen on this stream `.ga_data` is `NULL`.

oni-link commented on src/nvim/eval.c:
```diff
@@ -22397,318 +22342,54 @@ char_u *do_string_sub(char_u *str, char_u *pat, char_u *sub,
+      on_stdout->self = vopts;
+    }
+    if (on_stderr->buffered && on_stderr->cb.type == kCallbackNone) {
+      on_stderr->self = vopts;
```
Can `self` still be accessed if all callbacks have type `kCallbackNone`?

bfredl replied:
yes, buffered output is then stored in `self`.

oni-link commented on runtime/doc/channel.txt:
```diff
@@ -0,0 +1,172 @@
+	  - `abc\nefg` may arrive as `['abc', '']`, `['efg']` or `['abc']`,
+	    `['','efg']`, or even `['ab']`, `['c','efg']`.
+	
+	If you only are interesented in complete output when process exits,
```
interesented -> interested

oni-link replied:
when process -> when the process

oni-link commented on runtime/doc/channel.txt:
```diff
@@ -0,0 +1,172 @@
+	    `['','efg']`, or even `['ab']`, `['c','efg']`.
+	
+	If you only are interesented in complete output when process exits,
+	use buffered mode. Otherwise, and easy way to deal with this:
```
and -> an

bfredl commented:
@justinmk I would suggest getting this merged early in the 0.2.2 cycle, so it can "brew on master" for a bit. (and enable dependent PRs, including UpdateRemotePlugins refactor). Docs certainly can use more polish, but those can be improved after hand. After all, all existing functions only passing existing options as of 0.2.1 should behave the same with this PR.

bfredl commented:
Also last commit prototypes changing `b:terminal_channel_id` to buffer local option `'channel'` as was suggested in comment (but I can't find it now).

oni-link commented on src/nvim/buffer_defs.h:
```diff
@@ -603,6 +603,7 @@ struct file_buffer {
   char_u *b_p_bt;               ///< 'buftype'
   int b_has_qf_entry;           ///< quickfix exists for buffer
   int b_p_bl;                   ///< 'buflisted'
+  long b_p_channel;             ///< 'bomb'
```
Copy paste error: `bomb`? Comment should be `channel`.

oni-link replied:
Do we need a bigger type for the channel number or is `INT_MAX` an upper limit that we don't reach?

bfredl replied:
#6917 will make this limit explicit.

oni-link commented on src/nvim/channel.c:
```diff
@@ -706,6 +706,7 @@ void channel_terminal_open(Channel *chan)
   topts.resize_cb = term_resize;
   topts.close_cb = term_close;
   Terminal *term = terminal_open(topts);
+  curbuf->b_p_channel = (long)chan->id;
```
Should this option be set in `terminal_open()`, before event `TermOpen` is applied? 
Does the last section in the help for `terminal-status` needs to be updated after the channel id became an option?

```
These variables are initialized before TermOpen, so you can use them in
a local 'statusline'. Example: >
    :autocmd TermOpen * setlocal statusline=%{b:term_title}
```

bfredl replied:
sounds like a good idea.

oni-link commented on src/nvim/eval.c:
```diff
@@ -16710,9 +16710,6 @@ static void f_termopen(typval_T *argvars, typval_T *rettv, FunPtr fptr)
-  // TODO(bfredl): maybe read-only option instead?
-  dict_set_var(curbuf->b_vars, cstr_as_string("terminal_channel_id"),
-               INTEGER_OBJ(chan->id), false, false, &err);
   api_clear_error(&err);
```
This function call can also be removed if the dictionary variable is not set.

justinmk commented on test/functional/provider/ruby_spec.lua:
```diff
@@ -38,7 +38,7 @@ describe(':ruby command', function()
   end)
 
-  it('supports nesting', function()
+  pending('supports nesting', function()
```
Haven't seen this issue for several days so it might be safe to remove this.

bfredl committed:
provider: delete vimL stderr collector, now that it exists builtin

bfredl commented:
@justinmk If you (or anyone else) won't protest I would merge this. Docs a still a bit rough, especially fb3c21e09070b87b1ac734be14109c0e84cd9b4f is not fully integrated, but doc work will continue in #6743 which will fully formalize the terminology of channel types. Again, no behavior of any existing valid function call should have changed (except for the bugfix of making event handling of job rpc channels work like other rpc channels)

bfredl commented:
merged. The "channels can connect to socket" (bytes mode) test was a bit flaky the other day, but has been fine the latest retries. If it causes trouble again I will mark it as pending (and eventually fix it by adding a proper bytes server shim, instead of abusing msgpackdump for this purpose).

lambdalisue commented:
This seems breaking old behavior... An extra newline seems added. I'll investigate deeply later.

https://travis-ci.org/lambdalisue/gina.vim/jobs/308500736#L2077
https://ci.appveyor.com/project/lambdalisue/gina-vim/build/655/job/brq82xko9xudoi8p#L115

bfredl commented:
Hmm, I first though it was issue of the extra EOF event that is added. But `gina#util#extend_content()` looks like it should handle it perfectly fine. ((unrelated note: you can avoid the extra branch by initializing `job.stdout/stderr` with `['']` instead of `[]`) Can you confirm it is what's been used in this test and not some other stuff, in vital for instance?

lambdalisue commented:
https://travis-ci.org/lambdalisue/gina.vim/builds/308500727

Only the neovim HEAD shows fail so I think test is correct.

lambdalisue commented:
> I first though it was issue of the extra EOF event that is added.

Do you have detail? I would like to investigate that.

> Can you confirm it is what's been used in this test and not some other stuff, in vital for instance?

Possible. I'll check thanks 👍

lambdalisue commented:
https://github.com/lambdalisue/gina.vim/blob/master/test/gina/process.vimspec#L31

You were correct. The test has its own inplementation which just add the data to a list.

So is this behavior change intentional? And is there any documentation about behavior change?

justinmk commented:
@lambdalisue See "Note 2" in `:help job_control`. The behavior didn't change, it may have worked by chance before.

wsdjeg commented:
> 90 % of plugins that uses the recommended pattern should notice no breaking change

so what is the recommanded pattern ?

if a `shell commnd` output in terminal is:

```
fooo

foo
foo
```

then the a:data of stdout func is `['foo', '', 'foo', 'foo']`?

and any document show the differences, or any job example show the differences?

lambdalisue commented:
> 90 % of plugins that uses the recommended pattern should notice no breaking change

In my case, I was using a low-level implementation (only in one test, other parts I was using a recommended pattern) so the test had failed.
Once I changed the implementation as an example in a help, the test pass :tada:

https://github.com/lambdalisue/gina.vim/commit/0bdb8445ec084d1580b124ba46fdb2b91c4eb813

@wsdjeg 

Prob "Note 2" in `:help job_control` is the recommended pattern.

wsdjeg commented:
@lambdalisue Thanks, I got it.

wsdjeg commented:
@bfredl is there a check flag for this change? just to make sure my config is compatibility to old version of neovim, for example

```vim
   " some new func added in this pr
   if exist('*new_func_name')
       " old stdout handle func
   else
      " new fixed stdout handle func
   endif
```

bfredl commented:
The point is that a function that parses lines correctly,  _shouldn't_ need to change behavior depending on version. But you can of course check for `chansend()` existing if really want.

justinmk commented on src/nvim/os/shell.c:
```diff
@@ -231,32 +227,29 @@ static int do_os_system(char **argv,
-  // process events, so reset everything to NULL. It prevents closing the
+  // Note: unlike process events, stream events are not queued, as we want to
+  // deal with stream events as fast a possible.  It prevents closing the
   // streams while there's still data in the OS buffer (due to the process
```
Does "It prevents" still make sense here? What prevents the streams from closing?

boxofrox commented:
Are there any plans to extend `sockconnect` to work without the RPC flag?  I found a use-case for such a feature while working with [vim-picker](https://github.com/srstevenson/vim-picker).

bfredl commented:
@boxofrox This is already implemented, but only on master.
